File: leetcode/reverse_linked_list/reverse_linked_list.cpp
/*
 * Time Complexity: O(N)
 *   - N is the number of nodes in the linked list.
 *   - Each node is visited once.
 *
 * Space Complexity: O(1)
 *   - Only a constant amount of extra space is used.
 */

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode *p, *tmp;
        while(head != nullptr) {
            tmp = head->next;
            head->next = p;
            p = head;
            head = tmp;
        }
        return p;
    }
};

int main() {
    ListNode list(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
    Solution s;
    s.reverseList(&list);
}

File: leetcode/reverse_linked_list/reverse_linked_list.java
package reverse_linked_list;

class ListNode {
   int val;
   ListNode next;
   ListNode() {}
   ListNode(int val) { this.val = val; }
   ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution {
    public ListNode reverseList(ListNode head) {
        if(head == null) return head;
        ListNode previous = null, tmp = null;

        while(head != null) {
            tmp = head.next;
            head.next = previous;
            previous = head;
            head = tmp;
        }

        return previous;
    }
}
